library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileInputOperator.java

```diff
@@ -569,9 +601,18 @@ protected void replay(long windowId)
       Map<Integer, Object> recoveryDataPerOperator = windowDataManager.retrieveAllPartitions(windowId);
 
       for (Object recovery : recoveryDataPerOperator.values()) {
-        @SuppressWarnings("unchecked")
-        LinkedList<RecoveryEntry> recoveryData = (LinkedList<RecoveryEntry>)recovery;
 
+        WindowRecoveryEntry windowRecoveryEntry = (WindowRecoveryEntry)recovery;
+        List<RecoveryEntry> recoveryData = windowRecoveryEntry.fileRecoveryInfo;
+
+        if (windowRecoveryEntry.startBatchEmitted) {
+          emitStartBatchControlTuple();
+          return;
+        }
+        if (windowRecoveryEntry.endBatchEmitted) {
```

Comment:
Note that one window will emit only one among StartBatch and EndBatch control tuples.